src/util.c

```diff
@@ -132,44 +132,314 @@ long PURE str_to_long(const char *str)
+// number which can be used for an unconditional shift. If the upper (resp.
+// lower) half is all zeros, the test produces a zero, and the shift is a
+// no-op. A branchless implementation has the disadvantage that it requires
+// more instructions to execute than one which branches, but the advantage is
+// that none will be mispredicted branches. Whether this is a good tradeoff
+// depends on the branch predictability and the architecture's pipeline depth.
+// The most critical use of clzl in the kernel is in the scheduler priority
+// queue. In the absence of a concrete application and hardware implementation
+// to evaluate the tradeoff, we somewhat arbitrarily choose a branchless
+// implementation. In any case, the compiler might convert this to a branching
+// binary.
+
+// Check some assumptions made by the CLZ and CTZ library functions:
+compile_assert(clz_ulong_32_or_64, sizeof(unsigned long) == 4 || sizeof(unsigned long) == 8);
+compile_assert(clz_ullong_64, sizeof(unsigned long long) == 8);
+
+#ifdef CONFIG_CLZL_IMPL
+// Count leading zeros.
+// This implementation contains no branches. If the architecture provides an
+// instruction to set a register to a boolean value on a comparison, then the
+// binary might also avoid branching. A branchless implementation might be
+// preferable on architectures with deep pipelines, or when the maximum
+// priority of runnable threads frequently varies. However, note that the
+// compiler may choose to convert this to a branching implementation.
+static CONST inline unsigned clz32(uint32_t x)
 {
-    uint32_t count = 0;
-    while (!(x & 0x80000000U) && count < 34) {
-        x <<= 1;
-        count++;
+    // Compiler builtins typically return int, but we use unsigned internally
+    // to reduce the number of guards we see in the proofs.
+    unsigned count = 32;
+    uint32_t mask = UINT32_MAX;
+
+    // Each iteration i (counting backwards) considers the least significant
+    // 2^(i+1) bits of x as the current focus. At the first iteration, the
+    // focus is the whole input. Each iteration assumes x contains no 1 bits
+    // outside its focus. The iteration contains a test which determines
+    // whether there are any 1 bits in the upper half (2^i bits) of the focus,
+    // setting `bits` to 2^i if there are, or zero if not. Shifting by `bits`
+    // then narrows the focus to the lower 2^i bits and satisfies the
+    // assumption for the next iteration. After the final iteration, the focus
+    // is just the least significant bit, and the most significsnt 1 bit of the
+    // original input (if any) has been shifted into this position. The leading
+    // zero count can be determined from the total shift.
+    //
+    // The iterations are given a very regular structure to facilitate proofs,
+    // while also generating reasonably efficient binary code.
+
+    // The `if (1)` blocks make it easier to reason by chunks in the proofs.
+    if (1) {
```

Comment:
Does this have to be `if (1) {` or could it just be just an anonymous block scope with `{` also?

Reply:
Hi @axel-h, sorry I only saw your comment as I was cleaning up branches minutes after merging this!

I wrote this code some months ago, and I can't recall if I tested what you're suggesting. It's quite possible that I didn't test it, and just assumed that without the `if(1)` the C parser wouldn't give me what I wanted. I just spun up a test to give it a try...

Reply:
@axel-h Indeed, the C parser ignores the structure of anonymous block scopes. `... { ... } ...` produces the same Isabelle model as `... ... ...`. So it seems that `if (1) { ... }` is the most concise way to get what I wanted.

But since you made me think about it some more... It occurs to me that extracting the block contents to a `static inline` function would likely produce the same compiled binary, but would make the C code clearer and the verification even easier. I'll look into that if/when I find some time...

Reply:
How do you handle this usually, if this something that is supposed to be improved in the parser then? It can't completely ignore a block scope, but it may simply figure out when it can ignore if effectively (line in this case).

Reply:
Sorry, my statement was a bit too broad. I haven't looked at the details of the C parser, but I doubt it completely ignores block scopes. I should have said that in this particular code, block scopes don't produce any structure in the Isabelle model that we can use directly to structure the proof. With or without the block scopes, the C parser produces the same linear sequence of statements. On the other hand, it faithfully reproduces the redundant `if (1)` structure, and we can use that to structure the proofs (but as I realised earlier, a call to a `static inline` function would be better). We could potentially improve the C parser to do more with block scopes, but for now it doesn't seem to do much.

Reply:
Ok, that additional explanation makes more sense. Seems to me, there is a fundamental gap here - it's not really surprising that some semantics just get lost. What you'd ideally want to have is an annotated version of the generated assembly code, that still keeps some semantic information, so it can be used. The current way to hack this is in is by enforcing the creation of some assembly code that encode the semantics. Calling a static inline function would do the same in the end.
Hypothetically, if there were several different types of "nop" that would be used to encode the semantics, would this also be an option? In the RISC-V groups they have been discussing adding "hint" instructions, see loosely related to this.

Reply:
We're talking about the Isabelle model of the *C* here. The C parser preserves the structure of the `if (1)`, which we can use to structure the proofs about the C code, which we need for functional correctness.

The compiler erases the `if (1)` so that the structure is not visible in the binary. IIRC, it even produces different instructions for different blocks, because the most efficient way to get a particular constant in a register depends on the value of the constant. And that's all fine for binary verification, which is automatic. But the functional correctness proof requires manual effort, so it's useful to have the structure there.